extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java

```diff
@@ -1676,4 +1700,88 @@ public void run()
+          final List<PartitionInfo> partitionInfoList = topics.get(ioConfig.getTopic());
+          lagComputingConsumer.assign(Lists.transform(partitionInfoList, new Function<PartitionInfo, TopicPartition>()
+          {
+            @Override
+            public TopicPartition apply(PartitionInfo input)
+            {
+              return new TopicPartition(ioConfig.getTopic(), input.partition());
+            }
+          }));
+          final Map<Integer, Long> offsetsResponse = new HashMap<>();
+          final List<ListenableFuture<Void>> futures = Lists.newArrayList();
+          for (TaskGroup taskGroup : taskGroups.values()) {
+            for (String taskId : taskGroup.taskIds()) {
+              futures.add(Futures.transform(
+                  taskClient.getCurrentOffsetsAsync(taskId, false),
+                  new Function<Map<Integer, Long>, Void>()
+                  {
+                    @Override
+                    public Void apply(Map<Integer, Long> taskResponse)
+                    {
+                      if (taskResponse != null) {
+                        for (final Map.Entry<Integer, Long> partitionOffsets : taskResponse.entrySet()) {
+                          offsetsResponse.compute(partitionOffsets.getKey(), new BiFunction<Integer, Long, Long>()
+                          {
+                            @Override
+                            public Long apply(Integer key, Long existingOffsetInMap)
+                            {
+                              // If existing value is null use the offset returned by task
+                              // otherwise use the max (makes sure max offset is taken from replicas)
+                              return existingOffsetInMap == null
+                                     ? partitionOffsets.getValue()
+                                     : Math.max(partitionOffsets.getValue(), existingOffsetInMap);
+                            }
+                          });
+                        }
+                      }
+                      return null;
+                    }
+                  })
+              );
+            }
+          }
+          // not using futureTimeoutInSeconds as its min value is 120 seconds and this metric is emitted every minute
+          Futures.successfulAsList(futures).get(30, TimeUnit.SECONDS);
+
+          // for each partition, seek to end to get the highest offset
+          // check the offsetsResponse map for the latest consumed offset
+          // if partition info not present in offsetsResponse then use lastCurrentOffsets map
+          // if not present there as well, fail the compute
+
+          long lag = 0;
+          for (PartitionInfo partitionInfo : partitionInfoList) {
+            final TopicPartition topicPartition = new TopicPartition(ioConfig.getTopic(), partitionInfo.partition());
+            lagComputingConsumer.seekToEnd();
```

Comment:
Might be better to either take this out of the for loop and call it once, or call it partition specific, i.e. seekToEnd(topicPartition), even though it evaluates lazily inside the Kafka consumer